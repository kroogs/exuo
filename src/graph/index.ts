/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

export * from './useGraph'

export * from './models/factories'
export * from './models/Graph'
export * from './models/Node'

export * from './views/EdgeList'
export * from './views/LabelEditor'
export * from './views/NodeActions'
export * from './views/NodeLayout'
export * from './views/NodeListItem'
export * from './views/NodeViewer'
