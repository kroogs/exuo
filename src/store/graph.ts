/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Justin Krueger */

import { v4 as uuid } from 'uuid'
import {
  types as t,
  SnapshotIn,
  SnapshotOrInstance,
  Instance,
  IAnyType,
  IAnyModelType,
  IAnyStateTreeNode,
} from 'mobx-state-tree'

// TODO
// improve TS types
// types from configuration
// custom default graph type
// getEdgeMapTypes()

type EdgeTypeFn = () => IAnyType

export const edgeMapFactory = (getEdgeType: EdgeTypeFn): IAnyModelType =>
  t
    .model('EdgeMap', {
      edgeMap: t.map(t.array(t.reference(t.late(getEdgeType)))),
    })
    .actions(self => ({
      hasEdge(tag: string, target: Instance<IAnyModelType>): boolean {
        return Boolean(self.edgeMap.get(tag)?.includes(target))
      },
    }))
    .actions(self => ({
      addEdge(tag: string, target: Instance<IAnyModelType>) {
        if (!self.edgeMap.has(tag)) {
          self.edgeMap.set(tag, [])
        }

        self.edgeMap.get(tag)?.push(target)
      },

      getEdgeTag(tag: string): Array<Instance<IAnyModelType>> {
        return self.edgeMap.get(tag) ?? []
      },

      removeEdge(tag: string, target: Instance<IAnyModelType>) {
        if (!self.hasEdge(tag, target)) {
          throw Error(
            `No edge target '${target}' for tag '${tag}' on '${self}'`,
          )
        }

        self.edgeMap.get(tag)?.remove(target)
      },
    }))

export const nodeFactory = (
  getEdgeType: EdgeTypeFn,
  composeModels: Array<IAnyModelType | EdgeTypeFn> = [],
): IAnyModelType =>
  t.compose(
    // @ts-ignore
    ...[
      edgeMapFactory,
      ...composeModels,
      t.model({
        id: t.identifier,
      }),
    ].map(item => (typeof item === 'function' ? item(getEdgeType) : item)),
  )

const Node = nodeFactory(() => Node)

// const NodeType = t.model('NodeType', {
//   name: t.identifier,
//   compose: t.maybe(t.array(t.late((): IAnyType => t.reference(NodeType)))),
//   props: t.array(t.array(t.string)),
// })

export function graphFactory(
  initialNodeModels: Record<string, IAnyModelType> = { Node },
): IAnyModelType {
  const modelCache = {
    ...initialNodeModels,
  }

  const modelValues = Object.values(modelCache)

  return modelCache.Node.props({
    nodesById: t.map(
      modelValues.length > 1 ? t.union(...modelValues) : modelCache.Node,
    ),
    // typesById: t.map(NodeType),
  })
    .named('Graph')
    .actions(self => {
      // const createNodeTypeModel = (typeName: string): void => {
      //   const typeModel = self.typesById.get(typeName)
      //   if (!typeModel) {
      //     throw Error(`No type named '${typeName}'`)
      //   }

      //   const typeModelProps = Object.fromEntries(typeModel.props)
      //   typeModelCache[typeModel.name] = t
      //     .compose(Node, t.model(typeModelProps))
      //     .named(typeModel.name)
      // }

      return {
        createNode(
          modelName: string,
          props: Omit<SnapshotIn<IAnyModelType>, 'id'>,
        ): Instance<IAnyModelType> {
          if (!modelCache[modelName]) {
            throw Error(`No model named '${modelName}'`)
          }

          const instance = modelCache[modelName].create({
            ...props,
            id: uuid(),
          })

          self.nodesById.put(instance)

          return instance
        },
      }
    })
}

export const Graph = graphFactory()
