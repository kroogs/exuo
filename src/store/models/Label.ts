/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Justin Krueger */

import { types as t } from 'mobx-state-tree'

export const Label = t
  .model('Label', {
    label: t.maybe(t.string),
  })
  .actions(self => ({
    setLabel(label: string) {
      self.label = label
    },
  }))

export default Label
