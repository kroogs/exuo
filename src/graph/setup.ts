/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>
 *
 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react'
import { Instance } from 'mobx-state-tree'
import { useObserver } from 'mobx-react-lite'

import { Graph } from './models'

export const graphContext = React.createContext<Instance<typeof Graph>>(null)
export const GraphProvider: React.FunctionComponent<{
  value?: Instance<typeof Graph>
}> = ({ children, value }) =>
  React.createElement(
    graphContext.Provider,
    { value: value ?? Graph.create() },
    children,
  )

export function useGraph<S>(selector: (s: Instance<typeof Graph>) => S): S {
  const graph = React.useContext(graphContext)

  if (graph) {
    return useObserver(() => selector(graph))
  }

  throw Error('Cannot use graph before setup')
}
