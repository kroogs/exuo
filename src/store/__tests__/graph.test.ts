/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Justin Krueger */

import { types as t, getSnapshot, IAnyModelType } from 'mobx-state-tree'
import { edgeMapFactory, nodeFactory, graphFactory } from '../graph'
import Label from '../models/Label'

const Item = t.compose(
  edgeMapFactory((): IAnyModelType => Item),
  t.model({ id: t.identifier }),
)
const Container = t.model({
  items: t.map(Item),
})

describe('edgeMapFactory', () => {
  test('#addEdge stores a tagged reference to a target', () => {
    const box = Container.create({
      items: {
        one: { id: 'one' },
        two: { id: 'two' },
      },
    })

    const one = box.items.get('one')
    const two = box.items.get('two')

    one.addEdge('next', two)
    two.addEdge('prev', one)

    expect(getSnapshot(box)).toStrictEqual({
      items: {
        one: { id: 'one', edgeMap: { next: ['two'] } },
        two: { id: 'two', edgeMap: { prev: ['one'] } },
      },
    })
  })

  test('#hasEdge returns a boolean indicating a stored tagged reference exists', () => {
    const box = Container.create({
      items: {
        one: { id: 'one', edgeMap: { next: ['two'] } },
        two: { id: 'two' },
      },
    })

    const one = box.items.get('one')
    const two = box.items.get('two')

    expect(one.hasEdge('next', two)).toBe(true)
    expect(two.hasEdge('prev', one)).toBe(false)
  })

  test('#getEdgeTag returns an array of instances by tag', () => {
    const box = Container.create({
      items: {
        one: { id: 'one', edgeMap: { next: ['one', 'two'] } },
        two: { id: 'two' },
      },
    })

    const one = box.items.get('one')
    const two = box.items.get('two')

    expect(one.getEdgeTag('next')).toStrictEqual([one, two])
  })

  describe('#removeTest', () => {
    test('removes the first stored tagged reference to a target', () => {
      const box = Container.create({
        items: {
          one: { id: 'one', edgeMap: { next: ['two'] } },
          two: { id: 'two', edgeMap: { prev: ['one'] } },
        },
      })

      const one = box.items.get('one')
      const two = box.items.get('two')

      one.removeEdge('next', two)

      expect(getSnapshot(box).items).toStrictEqual({
        one: { id: 'one', edgeMap: { next: [] } },
        two: { id: 'two', edgeMap: { prev: ['one'] } },
      })

      two.removeEdge('prev', one)

      expect(getSnapshot(box).items).toStrictEqual({
        one: { id: 'one', edgeMap: { next: [] } },
        two: { id: 'two', edgeMap: { prev: [] } },
      })
    })

    test("throws an exception when an edge ref isn't found", () => {
      const box = Container.create({
        items: {
          one: { id: 'one' },
        },
      })

      const one = box.items.get('one')

      expect(() => one.removeEdge('next', one)).toThrowErrorMatchingSnapshot()
    })
  })
})

describe('graphFactory', () => {
  test('works', () => {
    const Node = nodeFactory(() => t.union(Node, LabelNode))
    const LabelNode = nodeFactory(() => t.union(Node, LabelNode), [Label])
    const GraphModel = graphFactory({ Node, LabelNode })

    const graph = GraphModel.create({ id: 'root' })

    // graph.createNode([
    //   {
    //     name: 'Test',
    //     props: [['label', 'string']],
    //   },
    // ])

    const one = graph.createNode('Node', {})
    const two = graph.createNode('Node', {})

    two.addEdge('side', one)

    expect(graph.nodesById.get(two.id).edgeMap.get('side')?.[0]).toBe(one)

    const potato = graph.createNode('LabelNode', { label: 'potato' })
    one.addEdge('item', potato)

    expect(graph.nodesById.get(one.id).edgeMap.get('item')?.[0]).toBe(potato)

    potato.addEdge('inside', potato)
    expect(potato.edgeMap.get('inside')?.[0]).toBe(potato)
  })
})
