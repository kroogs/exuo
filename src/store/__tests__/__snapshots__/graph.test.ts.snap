// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edgeMapFactory #removeTest throws an exception when an edge ref isn't found 1`] = `"No edge target 'AnonymousModel@/items/one(id: one)' for tag 'next' on 'AnonymousModel@/items/one(id: one)'"`;
